pkg/reconcile/pipeline/handler/project/impl.go

```diff
@@ -52,6 +52,10 @@ func PreFlightCheck(mandatoryBindingKeys ...string) func(pipeline.Context) {
 
 func PostFlightCheck(ctx pipeline.Context) {
 	ctx.SetCondition(apis.Conditions().InjectionReady().Reason("ApplicationUpdated").Build())
+	if ctx.HasLabelSelector() {
```

Comment:
iirc, `context.Applications()` return error in case when label selector is used and no workload is found: 

https://github.com/redhat-developer/service-binding-operator/blob/master/pkg/reconcile/pipeline/context/impl.go#L255 

That error is checked in `PreFlightCheck`: 

https://github.com/redhat-developer/service-binding-operator/blob/master/pkg/reconcile/pipeline/context/impl.go#L255

iirc, retrying should happen even with the current code?

Reply:
The intent of this patch is to pick up workloads that have labels *after* we've reconciled the service binding.  We're triggering a retry here _not_ because we failed to find any workloads, but because we succeeded and want to periodically rebind to get informed of new workloads with the labels we're interested in.  This is why I've left the error condition blank and put this in `PostFlightCheck` (i.e. the end of our pipeline).

As far as I know, operator-sdk doesn't let us install new reconcilers (especially on arbitrary resource types), so ideally, we'd be using a watch of some kind in a background task of some kind to get these resources and debounce them as they come in.  This patch is more of a workaround until we can properly do these kind of watches.

> iirc, retrying should happen even with the current code?

Correct, it does.  That addresses the error case when no workloads were found.

This patch is handling a separate case: we succeeded, and we've bound workloads that match a label selector.  However, we weren't binding new workloads that also match, which is required by the spec for us to support.

Reply:
> We're triggering a retry here not because we failed to find any workloads, but because we succeeded and want to periodically rebind to get informed of new workloads with the labels we're interested in. 

I understand the intention, but not sure if this is going to work as wished in practice. Check this excellent [blog post](https://danielmangum.com/posts/controller-runtime-client-go-rate-limiting/) describing in details how backoff strategies work. Simply, very quickly, reconciliation periods will become very large and there is also upper bound for number of tries, so at some point the reconciliation will happen after a number of hours, and eventually stop, and newly created workloads will not be bound.

Reply:
> Simply, very quickly, reconciliation periods will become very large and there is also upper bound for number of tries, so at some point the reconciliation will happen after a number of hours, and eventually stop, and newly created workloads will not be bound.

The way I understand it, the worst we'll see for a reconciliation delay by default is 16:40, which is long but not _super_ unreasonable.  If we used separate parameters on the rate-limiting scheme, we could tailor this down to be more frequent (with slight performance penalty).  Something like this:

```go
// SetupWithManager sets up the controller with the Manager.
func (r *BindingReconciler) SetupWithManager(mgr ctrl.Manager) error {
	pipeline, err := r.PipelineProvider(mgr.GetConfig(), kubernetes.ResourceLookup(mgr.GetRESTMapper()))
	if err != nil {
		return err
	}
	r.pipeline = pipeline
	return ctrl.NewControllerManagedBy(mgr).
		For(r.ReconcilingObject()).
		WithEventFilter(predicate.GenerationChangedPredicate{}).
		WithOptions(
			controller.Options{
				MaxConcurrentReconciles: MaxConcurrentReconciles,
				RateLimiter: workqueue.NewMaxOfRateLimiter(
					&workqueue.BucketRateLimiter{Limiter: rate.NewLimiter(rate.Limit(10), 100)},
					// set max timeout to 5 minutes (down from default of 1000 seconds)
					workqueue.NewItemExponentialFailureRateLimiter(5*time.Millisecond, 5*time.Minute),
				),
			}).
		Complete(r)
}
```

I'm also not convinced we're going to run into the upper bound.  I think that bound is implemented [here](https://github.com/kubernetes/client-go/blob/20732a1bc198ab57de644af498fa75e73fa44c08/examples/workqueue/main.go#L103), and we only run into that bound if we were reporting errors.  I explicitly set `err = nil` on those cases, so I don't think that this will be a problem.

Reply:
> The way I understand it, the worst we'll see for a reconciliation delay by default is 16:40, which is long but not super unreasonable.

Well, we speak here about 16mins, and IMHO from user perspective this is very long, at least user need to be aware that this is the time when the newly added workload will get the bindings projected.

Reply:
I've configured re-reconciliation times as a tunable parameter with a default of 2 minutes (chosen rather arbitrarily).  Since re-reconciliation has a performance impact, making this tunable by a cluster administrator seemed to be the right way to go.

Reply:
I am not sure if cluster admin has any way to discover and tune that parameter at the installation time. OLM does not support that. Also, even if admins want to update the deployment later on, the changes will be reverted by OLM. IMHO validate this change against a large cluster with a lot of service binding (both those that can be bound and others that cannot) /cc @pmacik

Reply:
BTW, you can implement your own `RateLimiter` where you can decide based on service binding item how long to wait for the next reconciliation. This way service bindings that declare workloads via labels can have constant retry times and do not interfere with other faulty service bindings.

Reply:
Even if you want to make the re-reconciliation time as tunable parameter, I would not bake this into a controller CLI arg - how about that you can control that per service binding, by applying an annotation on a given binding at the creation time, something like:

```yaml
kind: ServiceBinding
metadata:
  annotations:
     servicebinding.io/re-reconcilation-delays: 2m
.
.
.
```

Anyhow, this PR should contain as well an update to UserDoc describing the new behavior.